cabal-install/Distribution/Client/FileStatusCache.hs

```diff
@@ -0,0 +1,619 @@
+
+import           Control.Monad
+import           Control.Applicative
+import           Control.Monad.IO.Class
+import           Control.Monad.Trans.State (StateT)
+import qualified Control.Monad.Trans.State as State
+import           Control.Monad.Except
+import           Control.Exception
+
+import           Distribution.Text
+import           Distribution.Compat.ReadP ((<++))
+import qualified Distribution.Compat.ReadP as ReadP
+import qualified Text.PrettyPrint as Disp
+
+import           Distribution.Client.Glob
+import           Distribution.Client.Utils (mergeBy, MergeResult(..))
+
+import           System.FilePath
+import           System.Directory
+import           System.IO
+import           System.IO.Error
+import           GHC.Generics (Generic)
+
+
+------------------------------------------------------------------------------
+-- Types for specifying files to monitor
+--
+
+
+-- | A description of a file (or set of files) to monitor for changes.
+--
+-- All file paths here are relative to a common directory (e.g. project root).
+--
+data MonitorFilePath =
+
+     -- | Monitor a single file for changes, based on its modification time.
+     -- The monitored file is considered to have changed if it no longer
+     -- exists or if its modification time has changed.
+     --
+     MonitorFile !FilePath
+
+     -- | Monitor a single file for changes, based on its modification time
+     -- and content hash. The monitored file is considered to have changed if
+     -- it no longer exists or if its modification time and content hash have
+     -- changed.
+     --
+   | MonitorFileHashed !FilePath
+
+     -- | Monitor a single non-existent file for changes. The monitored file
+     -- is considered to have changed if it exists.
+     --
+   | MonitorNonExistentFile !FilePath
+
+     -- | Monitor a set of files identified by a file glob. The monitored glob
+     -- is considered to have changed if the set of files matching the glob
+     -- changes (i.e. creations or deletions), or if the modification time and
+     -- content hash of any matching file has changed.
+     --
+   | MonitorFileGlob !FilePathGlob
+
+  deriving (Show, Generic)
+
+instance Binary MonitorFilePath
+
+-- | A file path specified by globbing
+--
+data FilePathGlob
+   = GlobDir  !Glob !FilePathGlob
+   | GlobFile !Glob
+  deriving (Eq, Show, Generic)
+
+instance Binary FilePathGlob
+
+
+monitorFileSearchPath :: [FilePath] -> FilePath -> [MonitorFilePath]
+monitorFileSearchPath notFoundAtPaths foundAtPath =
+    MonitorFile foundAtPath
+  : map MonitorNonExistentFile notFoundAtPaths
+
+monitorFileHashedSearchPath :: [FilePath] -> FilePath -> [MonitorFilePath]
+monitorFileHashedSearchPath notFoundAtPaths foundAtPath =
+    MonitorFileHashed foundAtPath
+  : map MonitorNonExistentFile notFoundAtPaths
+
+-- | A file used to store the state of a file monitor.
+--
+-- It is typed to ensure it's used at consistent types.
+--
+newtype FileMonitorCacheFile a b
```

Comment:
Will document. It's a key and value. If the key has not changed, you get back the cached value. If the key has changed, you have to run an action to produce a new value.